app/controllers/work_packages/shares_controller.rb

```diff
@@ -39,41 +39,77 @@ def index
   end
 
   def create
-    WorkPackageMembers::CreateOrUpdateService
+    any_currently_shared_users = find_shared_users.any?
+
+    @share = WorkPackageMembers::CreateOrUpdateService
       .new(user: current_user)
       .call(entity: @work_package,
             user_id: params[:member][:user_id],
-            role_ids: find_role_ids(params[:member][:role_id]))
+            role_ids: find_role_ids(params[:member][:role_id])).result
+
 
-    respond_with_update_modal
+    if any_currently_shared_users
+      respond_with_prepend_share
+    else
+      respond_with_replace_modal
+    end
   end
 
   def update
     WorkPackageMembers::UpdateService
       .new(user: current_user, model: @share)
       .call(role_ids: find_role_ids(params[:role_ids]))
 
-    respond_with_update_modal
+    head :no_content
   end
 
   def destroy
     WorkPackageMembers::DeleteService
       .new(user: current_user, model: @share)
       .call
 
-    respond_with_update_modal
+    if find_shared_users.none?
+      respond_with_replace_modal
+    else
+      respond_with_remove_share
+    end
   end
 
   private
 
-  def respond_with_update_modal
+  def respond_with_replace_modal
     replace_via_turbo_stream(
       component: WorkPackages::Share::ModalBodyComponent.new(work_package: @work_package)
     )
 
     respond_with_turbo_streams
   end
 
+  def respond_with_prepend_share
+    prepend_via_turbo_stream(
+      component: WorkPackages::Share::ShareRowComponent.new(share: @share),
+      target_component: WorkPackages::Share::ModalBodyComponent.new(work_package: @work_package)
+    )
+
+    update_via_turbo_stream(
+      component: WorkPackages::Share::ShareCounterComponent.new(count: @shared_users.size)
+    )
+
+    respond_with_turbo_streams
+  end
+
+  def respond_with_remove_share
+    remove_via_turbo_stream(
+      component: WorkPackages::Share::ShareRowComponent.new(share: @share)
+    )
+
+    update_via_turbo_stream(
+      component: WorkPackages::Share::ShareCounterComponent.new(count: @shared_users.size)
```

Comment:
This only works because the `@shared_users` variable is set as a side effect in a method called by the same caller before this here is called. 

Using `find_shared_users` would be less surprising. Same applies to [further up](https://github.com/opf/openproject/pull/13845/files#diff-1044ba07b8889f8ff65fb7b1338806f8909221c74d285aadcb659334a3ca169dR95)